ignition-server/controllers/image_file_cache.go

```diff
@@ -0,0 +1,145 @@
+package controllers
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/openshift/hypershift/support/releaseinfo/registryclient"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+var log = ctrl.Log.WithName("image-cache")
+
+func NewImageFileCache(workDir string) (*imageFileCache, error) {
+	cacheDir, err := os.MkdirTemp(workDir, "cache-*")
+	if err != nil {
+		return nil, fmt.Errorf("failed to create cache directory: %w", err)
+	}
+
+	return &imageFileCache{
+		cacheMap:  make(map[cacheKey]cacheValue),
+		cacheDir:  cacheDir,
+		regClient: registryclient.ExtractImageFile,
+		RWMutex:   sync.RWMutex{},
+	}, nil
+}
+
+// Caching wrapper around the registry client function "regClient".
+// All files returned by the function are cached as files in a sub-directory of "cacheDir".
+// The cached files never expire, but can be automatically re-downloaded if they become corrupted.
+type imageFileCache struct {
+	cacheMap  map[cacheKey]cacheValue
+	cacheDir  string
+	regClient regClient // wrapped registryclient function
+	sync.RWMutex
+}
+
+type checksum []byte
+type regClient func(ctx context.Context, imageRef string, pullSecret []byte, file string, out io.Writer) error
+
+type cacheKey struct {
+	imageRef  string
+	imageFile string
+}
+
+type cacheValue struct {
+	fileName string
+	fileSha  checksum
+}
+
+func (c *imageFileCache) extractImageFile(ctx context.Context, imageRef string, pullSecret []byte, imageFile string, out io.Writer) error {
+	c.RLock()
+	defer c.RUnlock()
```

Comment:
I don't think we can just defer: I think we need to release and upgrade to a write lock at a point and only defer if the cache is hit